e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js

```diff
@@ -0,0 +1,189 @@
+
+const VIEWPORTS_SIDEBAR_IN_FLOW = [
+  [768, 1200],
+  [1024, 1200],
+  [1440, 1200],
+];
+
+describe("issue 31628", () => {
+  describe("display: scalar", () => {
+    beforeEach(() => {
+      restore();
+      cy.signInAsAdmin();
+      setupDashboardWithQuestionInCards(SCALAR_QUESTION, [
+        ...CARDS,
+        ...CARDS_SIZE_1X,
+      ]);
+    });
+
+    it("should render descendants of a 'scalar' without overflowing it (metabase#31628)", () => {
+      const descendantsSelector = [
+        "[data-testid='scalar-value']",
+        "[data-testid='scalar-title']",
+        "[data-testid='scalar-description']",
+      ].join(",");
+
+      VIEWPORTS_SIDEBAR_FLOATING.forEach(([width, height]) => {
+        cy.viewport(width, height);
+        assertDescendantsNotOverflowDashcards(descendantsSelector);
+      });
+
+      VIEWPORTS_SIDEBAR_IN_FLOW.forEach(([width, height]) => {
+        cy.viewport(width, height);
+        assertDescendantsNotOverflowDashcards(descendantsSelector);
+        openNavigationSidebar();
+        assertDescendantsNotOverflowDashcards(descendantsSelector);
+        closeNavigationSidebar();
+      });
+    });
+  });
+  describe("display: smartscalar", () => {
+    beforeEach(() => {
+      restore();
+      cy.signInAsAdmin();
+      setupDashboardWithQuestionInCards(SMART_SCALAR_QUESTION, CARDS);
+    });
+
+    it("should render descendants of a 'smartscalar' without overflowing it (metabase#31628)", () => {
+      const descendantsSelector = [
+        "[data-testid='scalar-value']",
+        "[data-testid='scalar-title']",
+        "[data-testid='scalar-description']",
+        "[data-testid='scalar-previous-value']",
+      ].join(",");
+
+      VIEWPORTS_SIDEBAR_FLOATING.forEach(([width, height]) => {
+        cy.viewport(width, height);
+        assertDescendantsNotOverflowDashcards(descendantsSelector);
+      });
+
+      VIEWPORTS_SIDEBAR_IN_FLOW.forEach(([width, height]) => {
+        cy.viewport(width, height);
+        assertDescendantsNotOverflowDashcards(descendantsSelector);
+        openNavigationSidebar();
+        assertDescendantsNotOverflowDashcards(descendantsSelector);
+        closeNavigationSidebar();
+      });
+    });
+  });
+});
+
+const setupDashboardWithQuestionInCards = (question, cards) => {
+  cy.createDashboard().then(({ body: dashboard }) => {
+    cypressWaitAll(
+      cards.map(card => {
+        return cy.createQuestionAndAddToDashboard(question, dashboard.id, card);
+      }),
+    );
+
+    visitDashboard(dashboard.id);
+  });
+};
+
+const assertDescendantsNotOverflowDashcards = descendantsSelector => {
+  cy.findAllByTestId("dashcard").each(dashcard => {
+    const descendants = dashcard.find(descendantsSelector);
+
+    descendants.each((_index, descendant) => {
+      assertDescendantNotOverflowsContainer(descendant, dashcard[0]);
+    });
+  });
+};
+
+const assertDescendantNotOverflowsContainer = (descendant, container) => {
```

Comment:
This function is generic and could be used in other tests. Do you think I should make it a helper or a command?